zebra-network/src/address_book.rs

```diff
@@ -5,54 +5,73 @@ use std::{
     collections::{BTreeSet, HashMap},
     iter::Extend,
     net::SocketAddr,
+    sync::Arc,
 };
 
 use chrono::{DateTime, Utc};
 use tracing::Span;
 
-use crate::{
-    constants,
-    types::{MetaAddr, PeerServices},
-};
+use crate::{constants, types::MetaAddr, PeerConnectionState};
 
 /// A database of peers, their advertised services, and information on when they
 /// were last seen.
 #[derive(Debug)]
 pub struct AddressBook {
-    by_addr: HashMap<SocketAddr, (DateTime<Utc>, PeerServices)>,
-    by_time: BTreeSet<MetaAddr>,
+    /// Maps each peer address to its matching `MetaAddr`
+    by_addr: HashMap<SocketAddr, Arc<MetaAddr>>,
+
+    /// Stores peer addresses, in state then reconnection order.
+    ///
+    /// This order should not be used for reconnection attempts: use
+    /// [`reconnection_peers`] instead.
+    by_state: BTreeSet<Arc<MetaAddr>>,
+
+    /// The span for operations on this address book.
     span: Span,
 }
 
 #[allow(clippy::len_without_is_empty)]
 impl AddressBook {
     /// Construct an `AddressBook` with the given [`tracing::Span`].
     pub fn new(span: Span) -> AddressBook {
-        AddressBook {
+        let constructor_span = span.clone();
+        let _guard = constructor_span.enter();
+
+        let new_book = AddressBook {
             by_addr: HashMap::default(),
-            by_time: BTreeSet::default(),
+            by_state: BTreeSet::default(),
             span,
-        }
+        };
+
+        new_book.assert_consistency();
+        new_book.update_metrics();
+
+        new_book
     }
 
     /// Get the contents of `self` in random order with sanitized timestamps.
     pub fn sanitized(&self) -> Vec<MetaAddr> {
         use rand::seq::SliceRandom;
+        let _guard = self.span.enter();
         let mut peers = self.peers().map(MetaAddr::sanitize).collect::<Vec<_>>();
         peers.shuffle(&mut rand::thread_rng());
         peers
     }
 
     /// Check consistency of the address book invariants or panic, doing work
-    /// quadratic in the address book size.
-    #[cfg(test)]
+    /// linear in the address book size.
     fn assert_consistency(&self) {
-        for (a, (t, s)) in self.by_addr.iter() {
-            for meta in self.by_time.iter().filter(|meta| meta.addr == *a) {
-                if meta.last_seen != *t || meta.services != *s {
-                    panic!("meta {:?} is not {:?}, {:?}, {:?}", meta, a, t, s);
-                }
-            }
+        assert_eq!(
+            self.by_addr.len(),
+            self.by_state.len(),
+            "state-based and addresss-based indexes must have the same number of entries"
```

Comment:
```suggestion
            "state-based and address-based indexes must have the same number of entries"
```

Might want to double check that this typo isn't elsewhere in the file since it showed up in two places

Reply:
Fixed by deleting the state-based index 🤔